### ISSUE TITLE
Integrating with outlines

### ISSUE THREAD

**Issue by harryjulian** [marella/ctransformers#91]
I'd love to use ctransformers with the [outlines](https://github.com/normal-computing/outlines) library for constrained generation. I opened this [issue ](https://github.com/normal-computing/outlines/issues/225) about it on their repo.

In order to hack away at this integration, the forward pass and pure logit output of the model would need to be exposed in the ctransformers API -- having looked through `ctransformers/llm.py` it seems as though the `.sample` method which wraps the `.ctransformers_llm_sample` method from the C API is the only way to sample tokens from the ctransformers models in it's current state. 

Would there be any scope for adding a `.forward` or `.sample_logits` method? How would I go about implementing it if there were?

**Comment by marella** [comment]
You can get the logits using [`llm.logits`](https://github.com/marella/ctransformers#property-llmlogits) property. In order to implement `forward`, you can use the low-level [`llm.eval()`](https://github.com/marella/ctransformers#method-llmeval) method.
I have done some work on this in the past to make ctransformers a drop-in replacement for 🤗 transformers models (see https://github.com/marella/ctransformers/issues/13#issuecomment-1597662836) [Here](https://github.com/marella/ctransformers/blob/guidance/ctransformers/transformers.py) is the code for reference.
Recently I created a better version of it but haven't pushed it yet. Here is a sample code from the newer version:

```py
import torch
from transformers import PretrainedConfig, PreTrainedModel
from transformers.modeling_outputs import CausalLMOutput

class Model(PreTrainedModel):
    def __init__(self, config: PretrainedConfig, llm):
        config.vocab_size = llm.vocab_size
        config.eos_token_id = llm.eos_token_id
        config.pad_token_id = llm.eos_token_id
        super().__init__(config)
        self._llm = llm
        self._past = []

    def prepare_inputs_for_generation(
        self,
        input_ids,
        attention_mask=None,
        **kwargs,
    ):
        return {"input_ids": input_ids}

    def forward(
        self,
        input_ids=None,
        return_dict=None,
        **kwargs,
    ):
        llm = self._llm
        tokens = input_ids.flatten().tolist()
        n_past = len(self._past)
        if tokens[:n_past] == self._past:
            self._past = tokens
            tokens = tokens[n_past:]
        else:
            self._past = tokens
            llm.reset()
        llm.eval(tokens)
        logits = torch.tensor(llm.logits).reshape([1, 1, -1])
        if not return_dict:
            return (logits,)
        return CausalLMOutput(logits=logits)

    @property
    def device(self) -> torch.device:
        return torch.device("cpu")
```

It can be used as:

```py
from ctransformers import AutoModelForCausalLM

llm = AutoModelForCausalLM.from_pretrained(...)
model = Model(PretrainedConfig(), llm)
```

But the API is not finalized and may change by the time I release it.

**Comment by marella** [comment]
Looks like outlines `generate` needs a custom model and tokenizer object which can be created using `Transformers` and `TransformersTokenizer` classes. I recommend using the original HF tokenizer to simplify things:

```py
model = Model(PretrainedConfig(), llm) # ctransformers model

from outlines.models.transformers import Transformers, TransformersTokenizer

tokenizer=TransformersTokenizer("Open-Orca/OpenOrca-Platypus2-13B") # change based on the model you are using for ctransformers
model = Transformers(model=model, tokenizer=tokenizer) # outlines model

answer = generate.choice(model, ["Positive", "Negative"])(prompt)
```

**Comment by Ozennefr** [comment]
For outlines 0.0.9, the forward pass should return a `CausalLMOutputWithPast` object instead of a `CausalLMOutput`. A drop in replacement work although more tweaks should allow for significant speedups.
